Frame switching W3C Compliance

andrewnicols opened the pull request:
Frame switching W3C Compliance

This change updates the frame functionality to comply with the W3C
specification.

Valid values of the `id` parameter are:
- null
- a Number between 0 and 2^16
- an Object representing a Web element

When the value is null, set the current browsing context to the current
top-level browsing context

When the value is a valid Number, let window be the associated window of
the current browsing context’s active document.

When the value is an Object, let element be the result of trying to get
a known element by web element reference id.

Source: https://www.w3.org/TR/webdriver/#switch-to-frame


I have also included a second commit, shared with #681 to pass isW3cCompliant to the constructor as we do in other classes.

OndraM commented on lib/Remote/RemoteTargetLocator.php:
```diff
@@ -63,10 +68,20 @@ public function defaultContent()
+        if ($this->isW3cCompliant) {
+            if ($frame instanceof WebDriverElement) {
+                $id = [JsonWireCompat::WEB_DRIVER_ELEMENT_IDENTIFIER => $frame->getID()];
+            } else if (null === $frame) {
```
```suggestion
            } elseif ($frame === null) {
```

Please don't use yoda conditions, they are harder to read :).

andrewnicols commented:
Hi @OndraM ,

I've realised that we need to produce a better error message when a string is passed. Right now if you pass a string as $frame, then it's cast to an int, specifically 0, and passes.

Not sure whether we should add a new Exception here or not?

The error returned by W3C WebDriver is:
```
Facebook\WebDriver\Exception\UnrecognizedExceptionException: invalid argument: 'id' can not be string
```

Sadly we don't have any kind of InvalidArgument exception or similar.

What are your thoughts?

I've added some unit tests and detect the current `Facebook\WebDriver\Exception\UnrecognizedExceptionException` but should be easy to change.

OndraM commented:
@andrewnicols It seems "invalid argument" error is missing in W3C exceptions handling 🤔.

Could you in this pull request add unit tests using `UnrecognizedExceptionException` and `expectExceptionMessage`, and I will have look at the missing exception :)?

OndraM commented:
@andrewnicols See #686 with all the new exceptions :ok_hand: .

You can now rebase your branch on top of `w3c-exceptions` branch and change base of this PR to `w3c-exceptions` - then you could already use the new exception.

andrewnicols commented:
Thanks @OndraM ,

All rebased.
Just thinking - what's the best way of excluding non-W3C browsers from the tests which will only work on W3C browsers? I.e. it's valid to use a string with a JSONWire browser but not with W3C. Do we have any ability to use any `@group` notation?

OndraM commented:
@andrewnicols You can use `self::skipForJsonWireProtocol()` method to skip those tests.

andrewnicols commented:
Ah thanks. I’ll do that when I get home in a couple of hours.

andrewnicols commented:
Right, Unit tests added and skipping the string unit test for JSONWire protocol

andrewnicols commented:
Note: This is based on #682 as it uses the activeElement functionality in the unit tests.

andrewnicols commented:
Rebased on community.

andrewnicols commented:
Rebased again:
* fixed missing `use` for exception
* corrected missing `is_numeric` check
* changed `assertEquals` to `assertContains` for content of the frame checks

andrewnicols commented:
And force pushed one final time because I had the order of the `use` statements wrong.

I'm going to stop changing things now!

OndraM commented:
I rebased the branch - there were conflicts following merger of #682

OndraM commented on lib/Remote/RemoteTargetLocator.php:
```diff
@@ -68,10 +68,22 @@ public function defaultContent()
+        if ($this->isW3cCompliant) {
+            if ($frame instanceof WebDriverElement) {
+                $id = [JsonWireCompat::WEB_DRIVER_ELEMENT_IDENTIFIER => $frame->getID()];
+            } else if ($frame === null) {
```
```suggestion
            } elseif ($frame === null) {
```

OndraM commented on lib/Remote/RemoteTargetLocator.php:
```diff
@@ -68,10 +68,22 @@ public function defaultContent()
+                $id = [JsonWireCompat::WEB_DRIVER_ELEMENT_IDENTIFIER => $frame->getID()];
+            } else if ($frame === null) {
+                $id = null;
+            } else if (is_numeric($frame)) {
```
```suggestion
            } elseif (is_numeric($frame)) {
```

OndraM commented on lib/Remote/RemoteTargetLocator.php:
```diff
@@ -68,10 +68,22 @@ public function defaultContent()
      */
```
Null should be mentioned in `@param` - and also with explanation what each value means (WebDriverElement/null/numeric id)

OndraM commented on lib/Remote/RemoteTargetLocator.php:
```diff
@@ -68,10 +68,22 @@ public function defaultContent()
+            } else if (is_numeric($frame)) {
+                $id = (int) $frame;
+            } else {
+                $id = (string) $frame;
```
If W3C only allows $frame to be either `WebDriverElement`, `null` or numeric ID, why should we accept string (and send it to the driver)?

andrewnicols replied:
So do you think this should except? The only reason I didn’t do so was because I wasn’t sure what exception I should throw. I haven’t seen any/many examples where we throw an exception rather than translate an error into an exception.

Should we just throw the same exception that would be thrown anyway?

OndraM replied:
We can just throw `\InvalidArgumentException` (the PHP's InvalidArgumentException, not the one from WebDriver). We already do it elsewhere in similar case.

OndraM commented on tests/functional/RemoteTargetLocatorTest.php:
```diff
@@ -71,4 +72,86 @@ public function testActiveElement()
+            'floa 1' => [
+                1.0,
+                $otherChildFrame,
+            ],
```
I would prefer having datasets formatted on one row if they fits, like this:
```php
 'null' => [null, $parentFrame],
```

andrewnicols replied:
I've removed some of these test cases as they were invalid.
We should not accept string or floats and cast to int because the frame number comes from a Number and it is perfectly valid to have a frame whose name is a stringy int.

I've added new tests for stringy integers.

I also detected an existing bug with the JSONWire implementation where it did not support being passed null.
I've confirmed that the following are supported in JSONWire (https://github.com/SeleniumHQ/selenium/wiki/JsonWireProtocol#sessionsessionidframe):
> id - {string|number|null|WebElement JSON Object} Identifier for the frame to change focus to.

OndraM commented on tests/functional/RemoteTargetLocatorTest.php:
```diff
@@ -71,4 +72,86 @@ public function testActiveElement()
+
+    /**
+     * Data provider for frame().
+     */
```
The comment is unnecessary

andrewnicols commented:
It looks like there may be some unit test fails.
I may have misread the inplementation of the numeric frame IDs. The Webdriver spec is particularly unclear on the meaning of them (See https://github.com/w3c/webdriver/issues/1205 for example).

It’s late here (Perth, Australia), so I won’t be able to look til tomorrrow unless you find a better solution before then.

Thanks

OndraM commented on tests/functional/RemoteTargetLocatorTest.php:
```diff
@@ -71,4 +72,86 @@ public function testActiveElement()
+    /**
+     * Data provider for frame().
+     */
+    public function frameProvider()
```
I would prefer this name (I will later rename all other providers to similar naming scheme `provide...()`)

```suggestion
    public function provideFrame()
```

OndraM commented on tests/functional/RemoteTargetLocatorTest.php:
```diff
@@ -71,4 +72,86 @@ public function testActiveElement()
+     * @cover ::frame
+     * @dataProvider frameProvider
+     */
+    public function testValidFrames($frame, $content)
```
```suggestion
    public function testValidFrames($frame, $expectedContent)
```

OndraM commented on tests/functional/RemoteTargetLocatorTest.php:
```diff
@@ -71,4 +72,86 @@ public function testActiveElement()
+        $this->assertContains(
+            $content,
+            $this->driver->switchTo()->activeElement()->getText()
+        );
```
I suggest separating every part of the arrange/act/assert phase of the test by new line. In this case it means newline after eachline:
```
        $this->driver->get($this->getTestPageUrl('page_with_frame.html'));

        $this->driver->switchTo()->frame($frame);

        $this->assertContains(
            $content,
            $this->driver->switchTo()->activeElement()->getText()
        );
```

OndraM commented on tests/functional/RemoteTargetLocatorTest.php:
```diff
@@ -71,4 +72,86 @@ public function testActiveElement()
+        $this->driver->switchTo()->frame($frame);
+        $this->assertContains(
+            $content,
+            $this->driver->switchTo()->activeElement()->getText()
```
Do we need to use `switchTo()->activeElement()`? Wouldn't something like `$this->driver->getPageSource()` do the trick? (also in the other tests)

andrewnicols replied:
Switched to getPageSource.

andrewnicols commented:
I’ll make those style changes tomorrow. Is there a specific style guide to follow for future reference so I don’t waste your time in future?

OndraM commented:
@andrewnicols Something is covered by codestyle check... Some other rules are not yet, sorry :-/. I plan to improve to codestyle check to contain more rules - but this will probably came with 2.0.

OndraM commented:
Nice @andrewnicols! I've done few changes in the tests, especially removed the data providers, because for every dataset in the provided new browser instance must be started - which is very time expensive. So it is generally better to avoid data provides in functional tests.

Otherwise looks good to merge :+1: , I will do so after CI builds finises. Nice work!
